lib/fizzy/execute.cpp

```diff
@@ -524,6 +524,33 @@ inline uint64_t popcnt64(uint64_t value) noexcept
     return static_cast<uint64_t>(__builtin_popcountll(value));
 }
 
+template <typename T>
+inline T fceil(T value) noexcept
+{
+    static_assert(std::is_floating_point_v<T>);
+    if (std::isnan(value))
+        return std::numeric_limits<T>::quiet_NaN();
+    return std::ceil(value);
```

Comment:
From https://en.cppreference.com/w/cpp/numeric/math/ceil:

![Screenshot 2020-08-14 at 19 31 28](https://user-images.githubusercontent.com/20340/90281598-c287d880-de64-11ea-86fc-435490c3f671.png)

What about this? The same applies for `std::floor`, and `std::trunc`.

Reply:
Okay, the documentation just kept calling it "floating point exception" and I thought it is an actual exception, not just a flag.

Can you leave perhaps a comment above ceil/floor that `FE_INEXACT` has no bearing on the correctness of this?

Reply:
This is GCC documentation about the floating point exceptions: https://www.gnu.org/software/libc/manual/html_node/FP-Exceptions.html
They can trigger SIGFPE signal, but this is not default behavior.

Reply:
Apparently can be forcefully disabled with `feenableexcept(0)`, see https://en.cppreference.com/w/cpp/numeric/fenv:

![Screenshot 2020-08-17 at 10 52 10](https://user-images.githubusercontent.com/20340/90383190-b6c03000-e077-11ea-9f30-5b309bc66e69.png)

Reply:
And there's the `non-call-exceptions` but that apparently also catches SIGSEGV.

Reply:
I think eventually we could have an enter/leave function for setting up FP properties. "enter" in the beginning of `execute` and then "enter"/"leave" wrapping on host functions.

Depends on which is more likely: heavy use of ceil/floor/trunc/round or not, because these are concerned by the rounding settings and the exceptions.

Reply:
Alternatively using the builtins wouldn't relieve us from this stdlib madness?

Reply:
> Alternatively using the builtins wouldn't relieve us from this stdlib madness?

This is not portable and we are using them currently anyway. This is the implementation of `std::ceil()` in libstdc++:
```cpp
  inline _GLIBCXX_CONSTEXPR float
  ceil(float __x)
  { return __builtin_ceilf(__x); }
```

I think sticking to the C stdlib is fine. If someone wants to enable GNU's extension to trap on FP exceptions be my guest.

Reply:
> If someone wants to enable GNU's extension to trap on FP exceptions be my guest.

Behaviour should be documented, and I'm leaning towards we should mention these in the README (since we don't have a complete API doc yet).